lib/Report/ProofOfPlay.php

```diff
@@ -1228,4 +1251,115 @@ private function groupByTagType(string $tagType) : string
+                        $row['tag'] . '_' . $row['widgetId'] . '_' . $row['parentCampaignId'] . '_' . $row['type'];
+
+                    if (!isset($data[$key])) {
+                        // Since we already have the display group as the grouping option, we can remove the display info
+                        $row['display'] = null;
+                        $row['displayId'] = null;
+                        $row['displayGroupId'] = $dg->displayGroupId;
+                        $row['displayGroup'] = $dg->displayGroup;
+
+                        $data[$key] = $row;
+                    } else {
+                        $data[$key]['duration'] += $row['duration'];
+                        $data[$key]['numberPlays'] += $row['numberPlays'];
+                    }
+                }
+            }
+        }
+
+        return $data;
+    }
+
+    /**
+     * Group by tag in MongoDB
+     * @param array $rows
+     * @param string $tagsType
+     * @return array
+     */
+    private function groupByTagMongoDb(array $rows, string $tagsType) : array
+    {
+        $data = [];
+        $tags = $this->filterByTagType($tagsType);
+        $type = match ($tagsType) {
+            'media' => 'mediaId',
+            'layout' => 'layoutId',
+            'dg' => 'displayId',
+        };;
+
+        foreach ($rows as $row) {
+            foreach ($tags as $tag) {
+                if ($row[$type] == $tag['entityId']) {
+                    // Create a temporary key to group by multiple columns at once
+                    // and save memory instead of checking each column recursively
+                    $key = $tag['tagId'] . '_' . $row['layoutId'] . '_' . $row['mediaId'] . '_' .
+                        $row['tag'] . '_' . $row['widgetId'] . '_' . $row['parentCampaignId'] . '_' . $row['type'];
+
+                    if (!isset($data[$key])) {
+                        // Since we already have the tags as the grouping option, we can remove the display info
+                        $row['display'] = null;
+                        $row['displayId'] = null;
+                        $row['tagName'] = $tag['tag'];
+                        $row['tagId'] = $tag['tagId'];
+
+                        $data[$key] = $row;
+                    } else {
+                        $data[$key]['duration'] += $row['duration'];
+                        $data[$key]['numberPlays'] += $row['numberPlays'];
+                    }
+                }
+            }
+        }
+
+        return $data;
+    }
+
+    /**
+     * @param string $tagsType
+     * @return array
+     */
+    private function filterByTagType(string $tagsType): array
+    {
+        $tags = [];
+        $filter = match ($tagsType) {
+            'media' => 'Media',
+            'layout' => 'Layout',
+            'dg' => 'Display',
+        };
+
+        // What type of tags are looking for?
+        foreach ($this->tagFactory->query() as $tag) {
```

Comment:
This is always looking at all possible tags, which feels like it could be optimised, but I am not sure why we look for them all?